parse.py

```diff
@@ -0,0 +1,222 @@
+        
+        Used properly during parsing, this function can turn infix notation into postfix notation.
+    
+        Args:
+            tokens (list): Tokens from which first token must be taken.
+        """
+        self._symbol_stack.append(tokens[0])
+
+    def push_unary_operator(self: object, tokens: list) -> None:
+        """Processes unary operators applied to atom and appends appropriate operators to symbol stack.
+        
+        For now, we have only two possible unary operators ('-' and '+'). 
+        +' does nothing and '-' turns the atom negative in odd numbers.
+    
+        Args:
+            tokens (list): Tokens belonging to atom. Unary operators to apply to atom will be at start of list.
+        """
+        # We will track parity of unary '-'s to push the 'unary -' symbol onto the stack only once per atom and improve runtime.
+        is_negative = False
+        for e in tokens:
+            if e == '-':
+                is_negative = not is_negative
+            elif e == '+':
+                continue
+            # The first symbol found that is not a unary operator marks the end of the unary operators.
+            else:
+                break
+        if is_negative:
+            self._symbol_stack.append('unary -')
+
+    def __init__(self: object) -> None:
+        """Define grammar to be used by parser and parse actions to be used in constructing the symbol stack.
+        """
+        #Expressions
+        expr = Forward()
+        
+        # Operations
+        plus, minus, multiply, divide, mod = map(Literal, "+-*/%")
+        left_bracket, right_bracket = map(Suppress, "()")
+        addition_operation = plus | minus
+        multiplication_operation = multiply | divide | mod
+        power_operation = Literal("^")
+        factorial_operation = Literal("!")
+        
+        # Functions
+        def add_arg_count_to_tokens(tokens):
+            function_id = tokens.pop(0)
+            arg_count = len(tokens[0])
+            tokens.insert(0, (function_id, arg_count))
+        function_id = Word(alphas)
+        argument_list = delimitedList(Group(expr)) # Expressions must be in groups so that we can count them seperately
+        function = (function_id + left_bracket + Group(argument_list) + right_bracket).setParseAction(add_arg_count_to_tokens)
+        
+        # Values
+        e = CaselessKeyword("E")
+        pi = CaselessKeyword("PI")
+        number = Regex(r"[+-]?\d+(?:\.\d*)?(?:[eE][+-]?\d+)?")
+        value = (e | pi | number)
+        
+        # Expression grammars. Order of operations is determined here.
+        factor   = Forward()
+        atom     = (addition_operation[...] + ((function | value).setParseAction(self.push_first) | Group(left_bracket + expr + right_bracket))).setParseAction(self.push_unary_operator)
+        factor   << atom    + ((power_operation + factor) | factorial_operation).setParseAction(self.push_first)[...]
```

Comment:
Based on this the exponent and factorial have the same precedence however when testing the factorial gets higher precedence? Can you explain for me why it happen?

Reply:
I'm not sure why this would mean that exponents and factorials have the same precedence. The exponent of a power has to be a factor according to the definition above. The argument to a factorial can be an atom. Since factors must be parsed from atoms, this means that factorials will have higher precedence than powers.

Reply:
here: atom    + ((power_operation + factor) | factorial_operation). The atom can be either of these form atom^factor or atom!. Therefore, from my pov they have the same precedence .

Reply:
No because the rule states that exponents must be created from a factor power and that factorials only require an atom as an argument. Ergo, if you have atom^atom!, the factorial MUST be processed first because you cannot have a factor as the exponent in a power. Then you would have atom^factor at which point the power would be processed as well.